Automattic/wp-calypso
# Assembler: Redirect back to the previous screen after the checkout

Related to https://github.com/Automattic/wp-calypso/issues/80947, https://github.com/Automattic/wp-calypso/issues/80679#issuecomment-1691047715

## Proposed Changes

* We want to redirect people back to the previous screen after the checkout. For example, when the user is on the Confirmation screen and continues to upgrade their plan. After the checkout, they have to redirect back to the Confirmation screen again, and then proceed to finish the Assembler flow.
* Hence, this PR is focusing on
  * Use the `useNavigatorListener` hook to sync the current screen and its parameter to the query parameters
  * When people land on the Assembler screen, we will build the initial path based on those query parameters
  * Use the `location.isInitial` to determine whether to use the `goBack` function or not. The reason is if the people start from the `styles` screen or `confirmation` screen, etc, the `goBack` function doesn't work since they are already on the first screen. So, we will use the `goTo` function to forward people to the initial path of the previous screen.
* Also, this PR made a change to move the back button next to the WordPress logo as it might be easy to handle the `goBack` behavior when people start the Assembler screen from the non-initial screen (main).

## Testing Instructions



* Go to `/setup?siteSlug=<your_site>` on your free site
* Continue to Design Picker
* Pick `Design your own` button to go to the Assembler screen
* On the Assembler screen
  * Ensure you are still on the same screen after the refresh
  * Ensure you are able to go back to the previous screen if you're not starting from the main screen
  * Pick the custom styles and go to checkout.
  * Ensure you are able to redirect back to the previous screen instead of the main screen when you
    * complete the checkout
    * cancel the checkout
    * Go to PayPal and complete the checkout
    * Go to PayPal and cancel the checkout

## Pre-merge Checklist
